### PR TITLE
Support join on constant

### LINKED ISSUE #25578: Support `LEFT JOIN ... ON 1 = 1` expression
**Usage**
Postgres, MySQL, and Oracle support the expression `LEFT JOIN ... ON 1 = 1`, but ClickHouse does not.

It may be possible to replace `LEFT JOIN ... ON 1 = 1` with `CROSS JOIN`, but I want to use the query in the same form for lots of databases.

Please support the expression `LEFT JOIN ... ON 1 = 1`.

**Detailed SQL**
```SQL
CREATE TABLE test
(
    `name` Nullable(String),
    `id` int
)
ENGINE = Log();

insert into test values ('a', 1) ('b', 2) (null, 3);

select * from test t1 left join (select * from test ) t2 on 1 = 1   
-- DB::Exception: Not equi-join ON expression: 1 = 1. No columns in one of equality side.: While processing 1 = 1.
```

### PR DESCRIPTION
**Pull request by vdimir**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

Support `JOIN ON 1 = 1` that have CROSS JOIN semantic, close #25578.

_Note:_ Joining on false or null not supported, but it's supported in postgress. It requires a bit more changes, I'll add it to this PR or will create other one.

### CODE REVIEW

**Comment by vdimir** [comment]
I suppose https://github.com/ClickHouse/ClickHouse/pull/24420 should be merged first

**Comment by vdimir** [comment]
> Functional stateless tests (memory) — fail: 1, passed: 3242, skipped: 14
`01079_bad_alters_zookeeper_long` looks not related, but test it not flaky

> Integration tests (asan) — fail: 34, passed: 68, flaky: 0
> Integration tests (thread) — fail: 34, passed: 1088, flaky: 3
`test_storage_rabbitmq` broken in master

**Comment by vdimir** [comment]
Do not merge before https://github.com/ClickHouse/ClickHouse/pull/21320 because of conflicts

**Review by KochetovNicolai — APPROVED** [review]

**Comment by alexey-milovidov** [comment]
@mergifyio update

File: src/Interpreters/HashJoin.cpp [hunk 1]
```diff
@@ -446,6 +451,23 @@ class KeyGetterForDict
     std::vector<size_t> positions;
 };
 
+/// Dummy key getter, always find nothing, used for JOIN ON NULL
```
[the remark this answers is not in the document]
Comment (vdimir) [reply src/Interpreters/HashJoin.cpp:454]: What do you mean by _JOIN from pipeline_ ?

**Comment by ka1bi4** [comment]
Internal documentation ticket: DOCSUP-18489

**Pull request merged**